jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionContext.kt

```diff
@@ -0,0 +1,554 @@
+        )
+        if (!frame1.isAccepted && !frame2.isAccepted &&
+            frame2.isImmediatelyAfter(frame1)) {
+            /* If neither frame is being projected, and they have consecutive
+               picture IDs, we don't need to leave any gap. */
+            seqGap = 0
+        } else {
+            /* If the earlier frame wasn't projected, and we haven't seen its
+             * final packet, we know it has to consume at least one more sequence number. */
+            if (!frame1.isAccepted && !frame1.seenEndOfFrame && seqGap > 1) {
+                seqGap--
+            }
+            /* Similarly, if the later frame wasn't projected and we haven't seen
+             * its first packet. */if (!frame2.isAccepted && !frame2.seenStartOfFrame && seqGap > 1) {
+                seqGap--
+            }
+            if (!frame1.isAccepted && seqGap > 0) {
+                seqGap--
+            }
+        }
+        return seqGap
+    }
+
+    private fun frameIsNewSsrc(frame: Vp9Frame): Boolean {
+        val lastFrame = lastVp9FrameProjection.vp9Frame
+
+        return lastFrame == null || !frame.matchesSSRC(lastFrame)
+    }
+
+    /**
+     * Find the previous frame before the given one.
+     */
+    @Synchronized
+    private fun prevFrame(frame: Vp9Frame): Vp9Frame? {
+        val frameMap = vp9PictureMaps.get(frame.ssrc) ?: return null
+        return frameMap.prevFrame(frame)
+    }
+
+    /**
+     * Find the next frame after the given one.
+     */
+    @Synchronized
+    private fun nextFrame(frame: Vp9Frame): Vp9Frame? {
+        val frameMap = vp9PictureMaps.get(frame.ssrc) ?: return null
+        return frameMap.nextFrame(frame)
+    }
+
+    /**
+     * Find the previous accepted frame before the given one.
+     */
+    private fun findPrevAcceptedFrame(frame: Vp9Frame): Vp9Frame? {
+        val frameMap = vp9PictureMaps.get(frame.ssrc) ?: return null
+        return frameMap.findPrevAcceptedFrame(frame)
+    }
+
+    /**
+     * Find the next accepted frame after the given one.
+     */
+    private fun findNextAcceptedFrame(frame: Vp9Frame): Vp9Frame? {
+        val frameMap = vp9PictureMaps.get(frame.ssrc) ?: return null
+        return frameMap.findNextAcceptedFrame(frame)
+    }
+
+    /**
+     * Find a subsequent base-layer TL0 frame after the given frame
+     * @param frame The frame to query
+     * @return A subsequent base-layer TL0 frame, or null
+     */
+    private fun findNextBaseTl0(frame: Vp9Frame): Vp9Frame? {
+        val frameMap = vp9PictureMaps.get(frame.ssrc) ?: return null
+        return frameMap.findNextBaseTl0(frame)
+    }
+
+    /**
+     * For a frame that's been accepted by the quality filter, verify that
+     * it's decodable given the projection decisions about previous frames
+     * (in case the targetIndex has changed).
+     */
+    private fun checkDecodability(frame: Vp9Frame): Boolean {
+        /* TODO - use SS or flexible mode reference list */
```

Comment:
I'm pretty sure that Chrome's VP9 decoder is smart enough that we won't need to.  Adding it would save us from potentially sending some undecodable frames when we switch which temporal layers we're sending, however at the expense of maintaining a fair bit more state and machinery in the bridge.  Do you think it's worth it?

Reply:
If we've got some reasonable way of determining if this is happening/causing issues then I think it's fine to leave as-is.

Reply:
Or maybe I'm misunderstood the potential risk--just thinking about: if we're wrong about the frequency, how hard will it be to deduce this is the issue from the symptoms?